---
title: "6_Library_Regressed"
author: "Romuald Parmentier"
date: "2025-04-01"
output: html_document
---

```{r Prepare environment, message = F}

# Libraries
library(SingleCellExperiment)
library(dplyr)
library(scran)
library(scater)
library(dplyr)

# Functions and palettes
source("/scicore/home/wykopa75/GROUP/rparmentier/sc_RNAseq/Projects/Prostate/Xenograft_Models/Xenograft_Models_Custom_Functions.R")
source("/scicore/home/wykopa75/GROUP/rparmentier/sc_RNAseq/Projects/Prostate/Xenograft_Models/Xenograft_Models_Color_Palettes.R")

# Create output path (adapted)
out_path = create_exp_folder(
  organ = "Prostate",
  project = "Xenograft_Models",
  samples_ID = "All_PDOXs_PDOXOs",
  exp = "6.1_Cell_Cycle_Regressed"
)

```

# Prepare the data

```{r Load the files}

# Load combined rescaled sce object after normalization and rescaling
file_path = get_exp_file_path(
  organ = "Prostate",
  project = "Xenograft_Models",
  samples_ID = "All_PDOXs_PDOXOs",
  prev_exp = "6.0_Dimension_Reduction_No_Regression", 
  pattern = "sce_obj.rds")

sce_obj = readRDS(file_path)

```

# Batch correction

```{r}

chosen.hvgs = rowSubset(sce_obj, "hvg_top_n1000")

fmnn.out <- suppressWarnings(batchelor::fastMNN( 
  sce_obj,
  batch = sce_obj$cell_cycle_phase,
  subset.row = chosen.hvgs, # Correction perfrormed on top 1000 hvg genes only
  k = 300, # Number of nearest neighbors (higher the smoother the correction is)
  d = 20) )# Number of dimension to retain

#  Store the MNN Values in the reducedDim slot of the combined object
reducedDim(sce_obj, "MNN.1-20") <- reducedDim(fmnn.out)

```

```{r Visualize FMNN loadings}

# Export fmnn loadings
fmnn_loadings = as_tibble(rowData(fmnn.out)$"rotation")
rownames(fmnn_loadings) = rownames(fmnn.out)

# Make atiblle out of fmnn loadings
MNN_loadings_df = as_tibble(fmnn_loadings)
colnames(MNN_loadings_df) = paste0("MNN", 1:ncol(fmnn_loadings))
rownames(MNN_loadings_df) = rownames(fmnn_loadings)

# Reshape from wide to long format
MNN_loadings_df_long <- MNN_loadings_df %>%
  mutate(gene = rownames(MNN_loadings_df)) %>%
  tidyr::pivot_longer(cols = -gene, names_to = "MNNs", values_to = "Loading")

# Calculate absolute values and arrange in descending order
top_MNN_loadings <- MNN_loadings_df_long %>%
  mutate(abs_loadings = abs(Loading)) %>%
  group_by(MNNs) %>%
  arrange(desc(abs_loadings)) %>%
  slice_head(n = 20)   

###################################
### Plot 20 first MNNS loadings ####
###################################

plot_list = list()
MNN_id = 1

MNNs = paste0("MNN", 1:20)

for(MNN in MNNs){
  
  top_MNN_loadings_subset = top_MNN_loadings %>%
    dplyr::filter(MNNs == MNN) %>%
    mutate(gene = factor(gene, levels = gene[order(abs_loadings, decreasing = F)]))  # Set factor levels based on decreasing order of loadings
  
  # Create a bar plot for top fmnn loadings, separated by principal components
  plot = ggplot(top_MNN_loadings_subset, aes(x = Loading , y = gene)) +  # 'gene' should be the identifier for rows/features
    geom_col() +
    labs(title = paste0("Top 20 loadings for MNN", MNN_id),
         x = "Absolute Loading",
         y = "Gene") +
    theme_minimal() 
  
  plot_list[[MNN_id]] = plot
  MNN_id = MNN_id + 1
  
}


multiple_page_layout <- gridExtra::marrangeGrob(
  grobs = plot_list, 
  nrow = 3, ncol = 3)

ggsave(
  plot = multiple_page_layout,
  filename = paste0(out_path,time_stamp(),"plot_MNN_1-20_loadings.pdf"),
  device = "pdf",
  height = 12,
  width = 10
)

```

```{r Choosing MNNs according loadings }

### Check how much the two first PCs are capturing sequencing depth (If ~1, then remove it)
##########################################################################################

MNN = reducedDim(sce_obj, "MNN.1-20")

# Store depth as the sum of counts for each cell (2 = sum of rows)
depth <- apply(counts(sce_obj), 2, sum)

paste("Correlation between MNN1 and sequenging depth with the raw data is:", 
      round(abs(cor(MNN[,1], depth)), digits = 2))

paste("Correlation between MNN2 and sequenging depth with the raw data is:", 
      round(abs(cor(MNN[,2], depth)), digits = 2))

paste("Correlation between MNN3 and sequenging depth with the raw data is:", 
      round(abs(cor(MNN[,3], depth)), digits = 2))

paste("Correlation between MNN4 and sequenging depth with the raw data is:", 
      round(abs(cor(MNN[,4], depth)), digits = 2))

```

# Uniform Manifold Approximation and Projection (UMAP) 

```{r Run UMAP}

sce_obj <- runUMAP(
  sce_obj,
  dimred = "MNN.1-20",
  name = "UMAP_on_MNN.1-20",
  n_neighbors = 300,
  min_dist = 0.7)

# Set default UMAP reduced dim slot
reducedDim(sce_obj,"UMAP") = reducedDim(sce_obj,"UMAP_on_MNN.1-20")

```

# Visualization 

```{r Make a tibble of metadata}

# get metadata
md <- as_tibble(colData(sce_obj))

# get embeddings from 2 first PC
coords <- as_tibble(reducedDim(sce_obj, "UMAP_on_MNN.1-20"))[,1:2]
colnames(coords) = c("UMAP_1","UMAP_2")

# combine dataframes
md <- cbind(md, coords)

md$Clean_ID = paste0(md$Model_ID,"_",md$Model_Mouse, "_", md$Model_Passage, "_", md$Model_System)


```

```{r Facetted UMAP colored per library}

sample_id = 1
plot_list = list()

for (sample in unique(md$SampleGroup)) {
  
  md_other <- subset(md, SampleGroup != sample)
  md_sample <- subset(md, SampleGroup == sample)
  
  plot <- ggplot() +
    # Rasterise the first layer
    ggrastr::rasterise(
      geom_point(
        data = md_other,
        aes(x = UMAP_1, y = UMAP_2), 
        fill = "grey", alpha = 0.4, size = 0.9,shape = 21, color = "gray10", stroke = 0.1
      ), dpi = 200, scale = 0.5
    ) +
    # Rasterise the second layer
    ggrastr::rasterise(
      geom_point(
        data = md_sample,
        aes(x = UMAP_1, y = UMAP_2, fill = Clean_ID), 
        alpha = 0.8, size = 1.5, shape = 21, color = "gray10", stroke = 0.2 
      ), dpi = 200, scale = 0.6
    ) +
    ggtitle(label = sample,
            subtitle = "Batch Correction (FastMNN applied)" ) +
    theme_classic() +
    scale_fill_manual(values = pal.PDOX.PDOXO.updated) +
    guides(
      fill = "none",
      alpha = "none",
      size = "none"
    )
  
  plot_list[[sample_id]] = plot
  
  sample_id = sample_id + 1
  
}

multiple_page_layout <- gridExtra::marrangeGrob(
  grobs = plot_list, 
  nrow = 2, ncol = 1)

ggsave(
  plot = multiple_page_layout,
  filename = paste0(out_path,time_stamp(),"plots_UMAP_Facetted_Sample.pdf"),
  device = "pdf",
  width = 210,
  height = 297,
  unit = "mm")


```

```{r Single UMAP colored with SampleGroup}

# Calculate centroids for each cluster
centroids <- md %>%
  group_by(Clean_ID) %>%
  summarise(
    UMAP_1 = mean(UMAP_1),
    UMAP_2 = mean(UMAP_2),
    .groups = 'drop'
  )

# Move the labels abo ve the cluster for each of them

plot <- ggplot() +
  # Rasterise the first layer
  ggrastr::rasterise(
    geom_point(
      data = md,
      aes(x = UMAP_1, y = UMAP_2, fill = Clean_ID), 
      alpha = 0.7, size = 1.3, 
      shape = 21,
      color = "gray10",
      stroke = 0.2),  # Lighter stroke color (adjust as needed)),# Ensure shape supports fill and color
    dpi = 300, scale = 1) +
  scale_fill_manual(values = pal.PDOX.PDOXO.updated) +
  theme_classic() +
  # Add cluster numbers in the center of each cluster
  # geom_text(
  #   data = centroids,
  #   aes(x = UMAP_1, y = UMAP_2, label = Clean_ID, color = Clean_ID),
  #   size = 4,
  #   fontface = "bold"
  # ) +
  # scale_color_manual(values = pal.PDOX.PDOXO.updated) +
  guides(
    # fill = "none",
    alpha = "none",
    size = "none",
    # color = "none"
  )

ggsave(
  plot = plot,
  filename = paste0(out_path,time_stamp(),"plot_UMAP_Single_PDO_Line_ID.pdf"),
  device = "pdf",
  width = 10,
  height = 8,
)

```

```{r Single UMAP colored with cell cycle}

# Calculate centroids for each cluster
centroids <- md %>%
  group_by(Clean_ID) %>%
  summarise(
    UMAP_1 = mean(UMAP_1),
    UMAP_2 = mean(UMAP_2),
    .groups = 'drop'
  )

# Move the labels abo ve the cluster for each of them

plot <- ggplot() +
  # Rasterise the first layer
  ggrastr::rasterise(
    geom_point(
      data = md,
      aes(x = UMAP_1, y = UMAP_2, fill = cell_cycle_phase), 
      alpha = 0.7, size = 1.3, 
      shape = 21,
      color = "gray10",
      stroke = 0.2),  # Lighter stroke color (adjust as needed)),# Ensure shape supports fill and color
    dpi = 300, scale = 1) +
  scale_fill_manual(values = pal_cell_cycle_phase) +
  theme_classic() +
  # Add cluster numbers in the center of each cluster
  # geom_text(
  #   data = centroids,
  #   aes(x = UMAP_1, y = UMAP_2, label = Clean_ID, color = Clean_ID),
  #   size = 4,
  #   fontface = "bold"
  # ) +
  # scale_color_manual(values = pal.PDOX.PDOXO.updated) +
  guides(
    fill = "none",
    alpha = "none",
    size = "none",
    # color = "none"
  )

ggsave(
  plot = plot,
  filename = paste0(out_path,time_stamp(),"plot_UMAP_Single_PDO_Cell_Cycle.pdf"),
  device = "pdf",
  width = 10,
  height = 8,
)

```

# Export sce object

```{r export the objects}

# Export annotated sce object
saveRDS(
  object = sce_obj,
  file = paste0(out_path,time_stamp(), "sce_obj.rds"))

```

